x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -0,0 +1,219 @@
+                                                               Consumer<Exception> errorHandler) {
+                Runnable task = () -> {
+                    try {
+                        IndexShard primaryShard = followerGroup.getPrimary();
+                        TransportWriteAction.WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> result =
+                            TransportBulkShardOperationsAction.shardOperationOnPrimary(primaryShard.shardId(), operations,
+                                primaryShard, logger);
+                        List<Translog.Operation> replicaOps = result.replicaRequest().getOperations();
+                        for (IndexShard replicaShard : followerGroup.getReplicas()) {
+                            TransportBulkShardOperationsAction.applyTranslogOperations(replicaOps, replicaShard, Origin.REPLICA);
+                        }
+                        handler.accept(primaryShard.getGlobalCheckpoint());
+                    } catch (Exception e) {
+                        errorHandler.accept(e);
+                    }
+                };
+                Thread thread = new Thread(task);
+                thread.start();
+            }
+
+            @Override
+            protected void innerSendShardChangesRequest(long from, int maxOperationCount, Consumer<ShardChangesAction.Response> handler,
+                                                        Consumer<Exception> errorHandler) {
+                Runnable task = () -> {
+                    IndexShard indexShard = getRandomIndexShard(leaderGroup);
+                    long globalCheckpoint = indexShard.getGlobalCheckpoint();
+                    try {
+                        Translog.Operation[] ops = ShardChangesAction.getOperations(indexShard, globalCheckpoint, from, maxOperationCount,
+                            params.getMaxBatchSizeInBytes());
+                        // Hard code index metadata version, this is ok, as mapping updates are not tested here.
+                        handler.accept(new ShardChangesAction.Response(1L, globalCheckpoint, ops));
+                    } catch (Exception e) {
+                        errorHandler.accept(e);
+                    }
+                };
+                Thread thread = new Thread(task);
+                thread.start();
+            }
+
+            @Override
+            protected boolean isStopped() {
+                return stopped.get();
+            }
+
+            @Override
+            public void markAsCompleted() {
+                stopped.set(true);
+            }
+
+            @Override
+            public void markAsFailed(Exception e) {
+                stopped.set(true);
+            }
+
+            private IndexShard getRandomIndexShard(ReplicationGroup replicationGroup) {
+                List<IndexShard> indexShards = new ArrayList<>(replicationGroup.getReplicas());
+                indexShards.add(replicationGroup.getPrimary());
```

Comment:
I ran into test failures when write ops are read from replica IndexShard instances in [innerSendShardChangesRequest(...)](https://github.com/elastic/elasticsearch/pull/32024/files#diff-a070668173fc36725eb048e4f080d183R182) method.

This is a failure that can happen for both tests:

```
Expected: <82>
     but: was <81>
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
		at org.junit.Assert.assertThat(Assert.java:956)
		at org.junit.Assert.assertThat(Assert.java:923)
		at org.elasticsearch.index.replication.ESIndexLevelReplicationTestCase$ReplicationGroup.assertAllEqual(ESIndexLevelReplicationTestCase.java:389)
		at org.elasticsearch.xpack.ccr.action.ShardFollowTaskReplicationTests.lambda$testSimpleCcrReplication$0(ShardFollowTaskReplicationTests.java:54)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:768)
		... 38 more
```

It misses the last operation. Looking from the trace log the global checkpoint reported from the index shard is off by one, causing this error.

If write ops are read from primary IndexShard instance then both tests pass. I checked this by always returning the primary IndexShard instance in this method, after many iterations no failure occurs. However changing it to return a replica shard instance then the test almost always fails. I suspect there is some race condition in the test framework here, but I'm unsure and still looking into it. 

@bleskes @dnhatn Do you have an idea why this happens?

Reply:
@martijnvg After the last write operation, we might need to sync the global checkpoint manually  in the test `leaderGroup.syncGlobalCheckpoint();`. This is fired async in the production code.

Reply:
Thanks @dnhatn, this fixed the problem: 59cdbcdf7f087999a239d1a22dd396af66b3fa9a